changes in checksums/results and some fixes (fixes#107)

djarecka opened the pull request:
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Summary

- [x] removing results_dict and using a new method `checksum_states` to get all checksums

- [x] restarting connections after running workflows

  -  test for workflows that are run twice (and use results right away)

  - closes #107

## Checklist
<!--- Please, let us know if you need help-->
- [x] All tests passing
- [x] I have added tests to cover my changes
- [ ] I have updated documentation (if necessary)
- [x] My code follows the code style of this project 
(we are using `black`: you can `pip install pre-commit`, 
run `pre-commit install` in the `pydra` directory 
and `black` will be run automatically with each commit)

## Acknowledgment
- [x] I acknowledge that this contribution will be available under the Apache 2 license.

satra commented on pydra/engine/core.py:
```diff
@@ -260,7 +277,7 @@ def __call__(self, submitter=None, plugin=None, **kwargs):
 
     def _run(self, **kwargs):
-        self.inputs = dc.replace(self.inputs, **kwargs)
+        # self.inputs = dc.replace(self.inputs, **kwargs) don't need it?
```
this was there to ensure that inputs could be set via the run command (when it was not a private function) or as a function call. as long as this is the case within `__call__` it would be ok to remove this. i.e. a user should be able to set inputs during the call to a task (much like a python function). 

if `**kwargs` is not being used, it should be removed.

djarecka replied:
oh, i see, we have `kwargs` in the `__call__` that is passing to `_run`, but had no tests for it, so I couldn't find who is using this. will keep it and add a test

djarecka replied:
added [a test](https://github.com/nipype/pydra/pull/130/files#diff-871957728874435d337bd031b6bcb53cR216)

satra commented on pydra/engine/core.py:
```diff
@@ -522,6 +543,7 @@ def add(self, task):
         """adding a task to the workflow"""
         if not is_task(task):
             raise ValueError("Unknown workflow element: {!r}".format(task))
+        task._reset()
```
does this need to happen at every `add` call? or could this be done just prior to `run` or perhaps more preferably at the end of a `run`?

satra commented on pydra/engine/core.py:
```diff
@@ -558,7 +582,7 @@ def create_connections(self, task):
 
     async def _run(self, submitter=None, **kwargs):
-        self.inputs = dc.replace(self.inputs, **kwargs)
+        # self.inputs = dc.replace(self.inputs, **kwargs) don't need it?
```
same question as before.

djarecka replied:
we do not allow running a workflow without a submitter, so I believe this can't be used

satra commented:
overall looks good. the only question i have is whether the reset can be called at the end of a run by the submitter, instead of during `add`. if so it may also help with the checksum issue.

```python
checksum1  = task.checksum
result = task(param1=2, param2=3)
checksum2 = task.checksum
assert checksum1 == checksum2
```

djarecka commented:
@satra - I'm just working on moving the `reset` to `Submitter`, so running workflow twice doesn't create a new output_dir, but other tests just started failing... might be related to `asyncio`, will try to think about it a bit more

djarecka commented:
@satra - at the end I just moved `wf._reset` to the end of `Submitter.__call__`. Wanted to put in more specific place but was having some issues.

about comparing the  `checksum`, in addition to the issue with resetting the connections, there was one more thing - we added `_graph_checksums` to inputs, and this is usually created in the submitter when the graph has all the nodes and it can sort it.
I've added a line to `checksum` that allow to set this before running.
